models/protected_tag.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"strings"
+
+	"code.gitea.io/gitea/modules/base"
+	"code.gitea.io/gitea/modules/timeutil"
+
+	"github.com/gobwas/glob"
+)
+
+// ProtectedTag struct
+type ProtectedTag struct {
+	ID               int64 `xorm:"pk autoincr"`
+	RepoID           int64
+	NamePattern      string
+	NameGlob         glob.Glob `xorm:"-"`
+	WhitelistUserIDs []int64   `xorm:"JSON TEXT"`
+	WhitelistTeamIDs []int64   `xorm:"JSON TEXT"`
+
+	CreatedUnix timeutil.TimeStamp `xorm:"created"`
+	UpdatedUnix timeutil.TimeStamp `xorm:"updated"`
+}
+
+// InsertProtectedTag inserts a protected tag to database
+func InsertProtectedTag(pt *ProtectedTag) error {
+	_, err := x.Insert(pt)
+	return err
+}
+
+// UpdateProtectedTag updates the protected tag
+func UpdateProtectedTag(pt *ProtectedTag) error {
+	_, err := x.ID(pt.ID).AllCols().Update(pt)
+	return err
+}
+
+// DeleteProtectedTag deletes a protected tag by ID
+func DeleteProtectedTag(pt *ProtectedTag) error {
+	_, err := x.ID(pt.ID).Delete(&ProtectedTag{})
+	return err
+}
+
+// EnsureCompiledPattern returns if the branch is protected
+func (pt *ProtectedTag) EnsureCompiledPattern() error {
+	if pt.NameGlob != nil {
+		return nil
+	}
+
+	expr := strings.TrimSpace(pt.NamePattern)
+
+	var err error
+	pt.NameGlob, err = glob.Compile(expr)
+	return err
+}
+
+// IsUserAllowed returns true if the user is allowed to modify the tag
+func (pt *ProtectedTag) IsUserAllowed(userID int64) (bool, error) {
+	if base.Int64sContains(pt.WhitelistUserIDs, userID) {
+		return true, nil
+	}
+
+	if len(pt.WhitelistTeamIDs) == 0 {
+		return false, nil
+	}
+
+	in, err := IsUserInTeams(userID, pt.WhitelistTeamIDs)
+	if err != nil {
+		return false, err
+	}
+	return in, nil
+}
+
+// GetProtectedTags gets all protected tags
+func (repo *Repository) GetProtectedTags() ([]*ProtectedTag, error) {
+	tags := make([]*ProtectedTag, 0)
+	return tags, x.Find(&tags, &ProtectedTag{RepoID: repo.ID})
+}
+
```

Comment:
```suggestion
// GetProtectedTags gets all protected tags
func (repo *Repository) GetProtectedTags() ([]*ProtectedTag, error) {
	return GetProtectedTags(repo.ID)
}

// GetProtectedTags gets all protected tags for the provided RepoID
func GetProtectedTags(repoID int64) ([]*ProtectedTag, error) {
	tags := make([]*ProtectedTag, 0)
	return tags, x.Find(&tags, &ProtectedTag{RepoID: repoID})
}

```

Reply:
I did not add a requirement for the repo passing in. I moved `rel.LoadAttributes()` to the top of the method which gets called always, so no change here. `LoadAttributes()` checks if `r.Repo == nil` and then loads it from the database. My changes in the other files setting `Repo` to something non nil are just an improvement to not query the database here unnecessarily. I'm happy to add this method but it does not solve a problem.

Reply:
I see that you've just moved the loadattributes up from lower down in the call. 

Ok.

My reasoning is that we should try to avoid calling loadattributes unless we actually need to do it as it will pull potentially a lot of objects out of the db. In this case your gettagprotection call doesn't really require anything from the repo except its id so it doesn't really need to be a function on the repo except as a convenience - and therefore Loadattributes would not be necessary for it. However as the call actually needs to call Loadattributes anyway it's fine to call it.